File: SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/ThreadScheduler.java
package com.braintreepayments.api.sharedutils;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.RestrictTo;
import androidx.annotation.VisibleForTesting;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class ThreadScheduler implements Scheduler {

    private final Handler mainThreadHandler;
    private final ExecutorService backgroundThreadService;

    ThreadScheduler() {
        this(new Handler(Looper.getMainLooper()), Executors.newCachedThreadPool());
    }

    @VisibleForTesting
    ThreadScheduler(Handler mainThreadHandler, ExecutorService backgroundThreadPool) {
        this.mainThreadHandler = mainThreadHandler;
        this.backgroundThreadService = backgroundThreadPool;
    }

    public void runOnBackground(Runnable runnable) {
        backgroundThreadService.submit(runnable);
    }

    public void runOnMain(Runnable runnable) {
        mainThreadHandler.post(runnable);
    }
}

File: SharedUtils/src/main/java/com/braintreepayments/api/sharedutils/ThreadScheduler.kt
package com.braintreepayments.api.sharedutils

import android.os.Handler
import android.os.Looper
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

internal class ThreadScheduler(
    private val mainThreadHandler: Handler = Handler(Looper.getMainLooper()),
    private val backgroundThreadService: ExecutorService = Executors.newCachedThreadPool()
) : Scheduler {

    override fun runOnBackground(runnable: Runnable) {
        backgroundThreadService.submit(runnable)
    }

    override fun runOnMain(runnable: Runnable) {
        mainThreadHandler.post(runnable)
    }
}
